content/docs/rfcs/12/README.md

```diff
@@ -61,77 +63,151 @@ The following are not considered as part of the adversarial model:
-
-A `FilterRequest` contains an optional topic, zero or more content filters and
-a boolean signifying whether to subscribe or unsubscribe to the given filters.
-True signifies 'subscribe' and false signifies 'unsubscribe'.
-
-A node that sends the RPC with a filter request and `subscribe` set to 'true' 
-requests that the filter node SHOULD notify the light requesting node of messages
-matching this filter.
-
-A node that sends the RPC with a filter request and `subscribe` set to 'false'
-requests that the filter node SHOULD stop notifying the light requesting node
-of messages matching this filter if it is currently doing so.
-
-The filter matches when content filter and, optionally, a topic is matched.
-Content filter is matched when a `WakuMessage` `contentTopic` field is the same.
-
-A filter node SHOULD honor this request, though it MAY choose not to do so. If
-it chooses not to do so it MAY tell the light why. The mechanism for doing this
-is currently not specified. For notifying the light node a filter node sends a
-MessagePush message.
-
-Since such a filter node is doing extra work for a light node, it MAY also
-account for usage and be selective in how much service it provides. This
-mechanism is currently planned but underspecified.
-
-#### MessagePush
-
-A filter node that has received a filter request SHOULD push all messages that
-match this filter to a light node. These [`WakuMessage`'s](./waku-message.md) are likely to come from the
-`relay` protocol and be kept at the Node, but there MAY be other sources or
-protocols where this comes from. This is up to the consumer of the protocol.
-
-A filter node MUST NOT send a push message for messages that have not been
-requested via a FilterRequest.
-
-If a specific light node isn't connected to a filter node for some specific
-period of time (e.g. a TTL), then the filter node MAY choose to not push these
-messages to the node. This period is up to the consumer of the protocol and node
-implementation, though a reasonable default is one minute.
+## Filter-Subscribe
+
+A filter service node MUST support the _filter-subscribe_ protocol
+to allow filter clients to subscribe, modify, refresh and unsubscribe a desired set of filter criteria.
+The combination of different filter criteria for a specific filter client node is termed a "subscription".
+A filter client is interested in receiving messages matching the filter criteria in its registered subscriptions.
+It is RECOMMENDED that filter service nodes allow only one subscription per client.
+
+Since a filter service node is consuming resources to provide this service,
+it MAY account for usage and adapt its service provision to certain clients.
+An incentive mechanism is currently planned but underspecified.
+
+### Filter Subscribe Request
+
+A client node MUST send all filter requests in a `FilterSubscribeRequest` message.
+This request MUST contain a `request_id`.
+The `request_id` MUST be a uniquely generated string.
```

Comment:
What happens if an user sends two `FilterSubscribeRequest` with the same `request_id`? Would that generate an error? or would the latest request received overwrite the subscription parameters of the first one?

Reply:
The latest request received would overwrite the subscription parameters of the first. We suggest now that service nodes only keep one subscription per peer (i.e. per `peerId`) and simply service requests as they come in (without keeping any `requestId` state - just making sure responses use the same `requestId`).
Reusing a `request_id` is now a spec violation on the side of the client, so the client would have to deal with the ambiguity of two responses on the same `request_id` (what if one fail and one succeeds?). It could then use the _refresh_ workflow to ensure the desired filter is installed (by just repeating its SubscribeRequest).

Reply:
> Imo, it would be most consistent if the implementation simply dropped the second request with the same id.

True, but we've now removed `requestId` as some identifier for the subscription and we don't (want to) keep track of `requestId`s in the service node. `SUBSCRIBE` now simply means subscribing to the included filter criteria (not replacing/removing any existing subscription - just extending it). The `requestId`, from the service node perspective, now simply ties the response to the request and is discarded directly afterwards by the service node. If it receives a subsequent `SUBSCRIBE` with the same `requestId` it will simply try to honor it with no reference to previous `requestId`s.